pkg/gardenlet/controller/federatedseed/extensions/controllerinstallation_control.go

```diff
@@ -0,0 +1,131 @@
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type controllerInstallationControl struct {
+	k8sGardenClient kubernetes.Interface
+	seedClient      kubernetes.Interface
+	seedName        string
+	log             *logrus.Entry
+
+	lock              *sync.Mutex
+	requiredKindTypes map[string]sets.String
+}
+
+func (c *controllerInstallationControl) createControllerInstallationReconcileFunc(ctx context.Context, kind string, newListObjFunc func() runtime.Object) reconcile.Func {
+	return func(req reconcile.Request) (reconcile.Result, error) {
+		// Step 1: List all existing extension objects of the given kind, collect the types and store it in the
+		// requiredKindTypes map.
+
+		listObj := newListObjFunc()
+		if err := c.seedClient.Client().List(ctx, listObj); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		requiredTypes := sets.NewString()
+		if err := meta.EachListItem(listObj, func(o runtime.Object) error {
+			obj, err := extensions.Accessor(o)
+			if err != nil {
+				return err
+			}
+
+			requiredTypes.Insert(obj.GetExtensionSpec().GetExtensionType())
+			return nil
+		}); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		c.lock.Lock()
+		c.requiredKindTypes[kind] = requiredTypes
+		c.lock.Unlock()
+
+		// Step 2: List all existing controller installation objects for the seed cluster this controller is responsible
+		// for, check if at least one of the supported extension kind/types is in the requiredKindTypes map. If yes then
+		// set Required=true, otherwise set Required=false.
+
+		controllerInstallationList := &gardencorev1beta1.ControllerInstallationList{}
+		if err := c.k8sGardenClient.Client().List(ctx, controllerInstallationList); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		controllerInstallationRequired := map[*gardencorev1beta1.ControllerInstallation]bool{}
+
+		for _, obj := range controllerInstallationList.Items {
+			controllerInstallation := obj.DeepCopy()
+
+			if controllerInstallation.Spec.SeedRef.Name != c.seedName {
+				continue
+			}
+
+			controllerRegistration := &gardencorev1beta1.ControllerRegistration{}
+			if err := c.k8sGardenClient.Client().Get(ctx, kutil.Key(controllerInstallation.Spec.RegistrationRef.Name), controllerRegistration); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			controllerInstallationRequired[controllerInstallation] = false
+			for _, resource := range controllerRegistration.Spec.Resources {
+				if requiredTypes, ok := c.requiredKindTypes[resource.Kind]; ok && requiredTypes != nil && requiredTypes.Has(resource.Type) {
```

Comment:
Shouldn't we also lock reads from map so that they don't happen during concurrent writes (with RWMutex for reads/writes)?

Reply:
Yeah, probably it's better to have it, thanks for spotting. I added it now, PTAL.